### PR TITLE
MeiliSearch refactor introducing OBKV format

### LINKED ISSUE #1531: Use OBKV format
Sub issue of https://github.com/meilisearch/MeiliSearch/issues/1593

Following: https://github.com/meilisearch/transplant/issues/142

----

To avoid 2 validations for the same input, we need to use the OBKV format on MeiliSearch's side and on Milli's side.

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR refactor some multiple components of meilisearch, and introduce the obkv document format to meilisearch

- [x] Split meilisearch-http and meilisearch-lib
- [x] Replace `IndexActor` and `UuidResolver` with `IndexResolver`
- [x] Remove mentions to Actor
- [x] Remove Actor traits to simplify code
- [x] Integrate obkv document format
- [x] Remove `Data`
- [x] Restore all route
- [x] Replace `Box<dyn error>` with `anyhow::Error`
- [x] Introduce update file store
- [x] Update file store error handling
- [x] Fix dumps
- [x] Fix snapshots
- [x] Fix tests
- [x] Update module documentation
- [x] add csv suppport (feat @ManyTheFish #1729 )
- [x] add jsonl support
- [x] integrate geosearch (feat @irevoire #1725) 

partially implements #1691 and #1690. The error handling is very basic now, I will finish it in the next pr.

Some unit tests have been disabled, I will re-enable them ASAP, but they need a bit more work.

close #1531 

P.S: sorry for this monstrous PR :'(

### CODE REVIEW

**Comment by curquiza** [comment]
Also fixed #1531

File: meilisearch-http/src/lib.rs [hunk 1]
```diff
@@ -1,69 +1,103 @@
...
+    Ok(())
+}
+
+pub fn configure_data(config: &mut web::ServiceConfig, data: MeiliSearch, opt: &Opt) {
+    let http_payload_size_limit = opt.http_payload_size_limit.get_bytes() as usize;
     config
-        .app_data(web::Data::new(data.clone()))
         .app_data(data)
+        // TODO!: Why are we passing the data with two different things?
+        //.app_data(data)
```
Comment (irevoire) [inline meilisearch-http/src/lib.rs:100]:
```suggestion
```
I think we can delete this now?

File: meilisearch-http/src/routes/indexes/documents.rs [hunk 2]
```diff
@@ -58,8 +62,26 @@ pub fn configure(cfg: &mut web::ServiceConfig) {
...
+            // update documents routes
+            .route(
+                web::put()
+                    .guard(empty_application_type)
+                    .to(|| HttpResponse::UnsupportedMediaType()),
+            )
+            .route(web::put().guard(guard_json).to(update_documents_json))
+            .route(web::put().guard(guard_ndjson).to(update_documents_ndjson))
+            .route(web::put().guard(guard_csv).to(update_documents_csv))
+            .route(web::put().to(|| HttpResponse::UnsupportedMediaType()))
```
Comment (irevoire) [inline meilisearch-http/src/routes/indexes/documents.rs]: Pretty cool :+1:

File: meilisearch-http/src/routes/indexes/search.rs [hunk 3]
```diff
@@ -81,14 +79,42 @@ impl From<SearchQueryGet> for SearchQuery {
...
+                merge = false;
+            }
+        } else {
+            sort_parameters.push(current_sort.to_string());
+            merge = false;
+        }
+    }
+    sort_parameters
+}
+
```
Comment (irevoire) [inline meilisearch-http/src/routes/indexes/search.rs:107]: Since I can run the tests now I'll rewrite this function after my review :smile:

File: meilisearch-lib/src/document_formats.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+                Some(Ok(document))
+            }
+            Err(e) => Some(Err(DocumentFormatError::MalformedPayload(
+                Box::new(e),
+                PayloadType::Csv,
+            ))),
+        }
+    }
+}
```
Comment (irevoire) [inline meilisearch-lib/src/document_formats.rs:163]: Do you think it would be a good idea for another PR to move all this code to a separate crate?
So we can use it in `milli/http-ui` and `milli/benchmarks` instead of copy-pasting it everywhere

File: meilisearch-lib/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                    UserError::InvalidFacetsDistribution { .. } => Code::BadRequest,
+                    UserError::InvalidSortableAttribute { .. } => Code::Sort,
+                    UserError::CriterionError(_) => Code::InvalidRankingRule,
+                    UserError::InvalidGeoField { .. } => Code::InvalidGeoField,
+                    UserError::SortError(_) => Code::Sort,
+                }
+            }
+        }
+    }
+}
```
Comment (irevoire) [inline meilisearch-lib/src/error.rs:62]: Why is this code written once in `meilisearch-http/src/error.rs` and once in `meilisearch-lib/src/error.rs` ?
I think we should remove one of the two :thinking:

File: meilisearch-lib/src/document_formats.rs (hunk 4 above)
Comment (MarinPostma) [reply meilisearch-lib/src/document_formats.rs:163]: I think, ultimately, the whole document format should be moved to another crate

File: meilisearch-lib/src/index_controller/index_resolver/mod.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    pub async fn get_index(&self, uid: String) -> Result<Index> {
+        match self.index_uuid_store.get_uuid(uid).await? {
+            (name, Some(uuid)) => {
+                match self.index_store.get(uuid).await? {
+                    Some(index) => Ok(index),
+                    None => {
+                        // For some reason we got a uuid to an unexisting index, we return an error,
+                        // and remove the uuid from th uuid store.
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/index_resolver/mod.rs:146]:
```suggestion
                        // and remove the uuid from the uuid store.
```

File: meilisearch-lib/src/index_controller/mod.rs [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+}
+
+/// Parses the v1 version of the Asc ranking rules `asc(price)`and returns the field name.
+pub fn asc_ranking_rule(text: &str) -> Option<&str> {
+    text.split_once("asc(")
+        .and_then(|(_, tail)| tail.rsplit_once(")"))
+        .map(|(field, _)| field)
+}
+
+/// Parses the v1 version of the Desc ranking rules `asc(price)`and returns the field name.
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/mod.rs:499]:
```suggestion
/// Parses the v1 version of the Desc ranking rules `desc(price)`and returns the field name.
```

File: meilisearch-lib/src/index_controller/mod.rs [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+pub fn desc_ranking_rule(text: &str) -> Option<&str> {
+    text.split_once("desc(")
+        .and_then(|(_, tail)| tail.rsplit_once(")"))
+        .map(|(field, _)| field)
+}
+
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/mod.rs:505]: Where is this code used? Shouldn't it only be used in `dump/extractor/v1.rs`?

File: meilisearch-lib/src/index_controller/snapshot.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+//uuid_resolver,
+//update_handle,
+//Duration::from_millis(100),
+//snapshot_path.path().to_owned(),
+//"data.ms".to_string(),
+//);
+
+//let _ = timeout(Duration::from_millis(300), snapshot_service.run()).await;
+//}
+//}
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/snapshot.rs:257]: What is this?
We should uncomment it or delete it I guess but not let it commented :thinking:

File: meilisearch-lib/src/index_controller/update_file_store.rs [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        dst.push(&uuid_string);
+
+        let update_file = File::open(update_file_path)?;
+        let mut dst_file = NamedTempFile::new()?;
+        let mut document_reader = DocumentBatchReader::from_reader(update_file)?;
+
+        let mut document_buffer = Map::new();
+        // TODO: we need to find a way to do this more efficiently. (create a custom serializer
+        // for
+        // jsonl for example...)
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/update_file_store.rs:145]:
```suggestion
        // for jsonl for example...)
```

File: meilisearch-lib/src/index_controller/updates/store/mod.rs [hunk 11]
```diff
@@ -254,8 +269,8 @@ impl UpdateStore {
         Ok(meta)
     }
 
-    /// Push already processed update in the UpdateStore without triggering the notification
-    /// process. This is useful for the dumps.
+    // /// Push already processed update in the UpdateStore without triggering the notification
+    // /// process. This is useful for the dumps.
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/updates/store/mod.rs:273]:
```suggestion
    /// Push already processed update in the UpdateStore without triggering the notification
    /// process. This is useful for the dumps.
```

File: meilisearch-lib/src/index_controller/updates/store/mod.rs [hunk 12]
```diff
@@ -435,16 +436,16 @@ impl UpdateStore {
...
         while let Some(Ok(((_, uuid, _), pending))) = pendings.next() {
             if uuid == index_uuid {
-                let mut pending = pending.decode()?;
-                if let Some(update_uuid) = pending.content.take() {
-                    uuids_to_remove.push(update_uuid);
-                }
+                let mut _pending = pending.decode()?;
+                //if let Some(update_uuid) = pending.content.take() {
+                //uuids_to_remove.push(update_uuid);
+                //}
```
Comment (irevoire) [inline meilisearch-lib/src/index_controller/updates/store/mod.rs:448]: Why can we remove this?

**Review by irevoire — COMMENTED** [review]
code-wise I don't see any big red flag but since there is so much code I probably missed a lot of things :grimacing: 
I'll try to test this PR this afternoon

File: meilisearch-lib/src/index_controller/snapshot.rs (hunk 9 above)
Comment (MarinPostma) [reply meilisearch-lib/src/index_controller/snapshot.rs:257]: I commented them out for now, since they require rewrite, I'll do it later, as I mentionned in the pr message

File: meilisearch-lib/src/index_controller/updates/store/mod.rs (hunk 12 above)
Comment (MarinPostma) [reply meilisearch-lib/src/index_controller/updates/store/mod.rs:448]: I forgot to restore it, thank you!!

File: meilisearch-lib/src/index_controller/mod.rs (hunk 8 above)
Comment (MarinPostma) [reply meilisearch-lib/src/index_controller/mod.rs:505]: this is also used in index/dump.rs

File: meilisearch-lib/src/index_controller/snapshot.rs (hunk 9 above)
Comment (irevoire) [reply meilisearch-lib/src/index_controller/snapshot.rs:257]: Yes sorry I noticed later that a lot of tests were commented out and that it was written in the PR :+1:

**Review by irevoire — APPROVED** [review]
Everything I tested worked.
I think we can merge, but since a lot of tests were commented out, I think we'll need to do a bunch of other PRs later to fix the tests and then fix the subtle bugs hidden in the code :smile:
